DFE-Digital/manage-courses-api
# Refactor user model

### Context
This is prep work for adding the logging of "invite sent date" (PR to follow). branch: https://github.com/DFE-Digital/manage-courses-api/tree/invite-tracking

This can't be merged until the change has been made to not empty mc_users on import ( #71 ) otherwise the sign-in id and datestamps will be lost on the next import.

### Changes proposed in this pull request
* Merge user_log into mc_users - this removes the email based fk
* fixup service & tests to match
* up/down migrations

### Guidance to review

testing the migrations:

- `dotnet ef --startup-project ..\ManageCourses.Api database drop`
- bring up to just before data migration:
- `dotnet ef --startup-project ..\ManageCourses.Api database update UserLog-ExtendMcUser`
- run sql to get some test data:

```
delete from mc_user;
insert into mc_user (id, email) values
(1,'bob@example.org'), -- simple join on id, completed user log data, access request by id
(2,'wendy@example.org'), -- minimal userlog record
(3,'mickey@example.org'), -- email changed, userlog data
(4,'adrian@example.org'), -- fk on id lost, access request + userlog
(5,'frank@example.org'); -- no userlog record or access request

delete from user_log;
insert into user_log (first_login_date_utc, last_login_date_utc, sign_in_user_id, user_email, user_id, welcome_email_date_utc) values
('2000-12-01','2000-12-02','signinguidhereAAA','bob@example.org',        1,   '2000-12-04'),
('2001-12-01','2001-12-02',null,        'wendy@example.org',      2,   null),
('2002-12-01','2002-12-02','signinguidhereCCC','mickey808@example.org',  3,   '2002-12-04'),
('2003-12-01','2003-12-02','signinguidhereDDD','adrian@example.org',     null,'2003-12-04');

delete from access_request;
insert into access_request (request_date_utc, requester_id, status, requester_email) values
('2010-12-01',1,   1,'bob@example.org'),
('2012-12-01',3,   1,'mickey45@example.org'),
('2013-12-01',null,1,'adrian@example.org');

insert into mc_organisation (name, org_id) values ('someorg', 'wut?');
insert into mc_organisation_user (email, org_id) values ('bob@example.org','wut?');
```

- `dotnet ef --startup-project ..\ManageCourses.Api migrations add update`
- check contents of McUser has log data
- test down migrations (this is the one before this set of migrations):
- `dotnet ef --startup-project ..\ManageCourses.Api database update WelcomeEmailColumn_UserLog`
- check contents of McUser has been restored
